modules/auxiliary/gather/ldap_esc1_cert_finder.rb

```diff
@@ -0,0 +1,420 @@
+      end
+      if (value & 0x00040000) == 0x00040000
+        raw_string += 'WRITE_DAC|'
+      end
+      if (value & 0x00080000) == 0x00080000
+        raw_string += 'WRITE_OWNER|'
+      end
+    end
+    if (value & 0x00100000) == 0x00100000
+      raw_string += 'SYNCHRONIZE|'
+    end
+    if (value & 0x001F0000) == 0x001F0000
+      raw_string += 'STANDARD_RIGHTS_ALL|'
+    end
+    if (value & 0x0000FFFF) == 0x0000FFFF
+      raw_string += 'SPECIFIC_RIGHTS_ALL|'
+    end
+
+    # Generic Access Codes
+    if (value & 0x80000000) == 0x80000000
+      raw_string += 'GENERIC_READ|'
+    end
+    if (value & 0x40000000) == 0x40000000
+      raw_string += 'GENERIC_WRITE|'
+    end
+    if (value & 0x20000000) == 0x20000000
+      raw_string += 'GENERIC_EXECUTE|'
+    end
+    if (value & 0x10000000) == 0x10000000
+      raw_string += 'GENERIC_ALL|'
+    end
+    if (value & 0x2000000) == 0x2000000
+      raw_string += 'MAXIMUM_ALLOWED|' # Should never be set normally though.
+    end
+    if (value & 0x1000000) == 0x1000000
+      raw_string += 'ACCESS_SYSTEM_SECURITY|'
+    end
+
+    # MS-DTYP Specific Access Control Codes. Taken from https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/f4296d69-1c0f-491f-9587-a960b292d070
+    if (value & 0x001F01FF) == 0x001F01FF
+      raw_string += 'FILE_ALL_ACCESS|'
+    else
+      if (value & 0x001200A0) == 0x001200A0
+        raw_string += 'FILE_EXECUTE|'
+      end
+      if (value & 0x00120116) == 0x00120116
+        raw_string += 'FILE_WRITE|'
+      end
+      if (value & 0x00120089) == 0x00120089
+        raw_string += 'FILE_READ|'
+      end
+    end
+    if (value & 0x000F003F) == 0x000F003F
+      raw_string += 'KEY_ALL_ACCESS|'
+    else
+      if (value & 0x00020019) == 0x00020019
+        raw_string += 'KEY_READ|KEY_EXECUTE|'
+      end
+      if (value & 0x00020006) == 0x00020006
+        raw_string += 'KEY_WRITE|'
+      end
+    end
+    if (value & 0x00000100) == 0x00000100
+      raw_string += 'CONTROL_ACCESS|'
+    end
+    if (value & 0x00000080) == 0x00000080
+      raw_string += 'LIST_OBJECT|'
+    end
+    if (value & 0x00000040) == 0x00000040
+      raw_string += 'DELETE_TREE|'
+    end
+    if (value & 0x00000020) == 0x00000020
+      raw_string += 'WRITE_PROPERTY|'
+    end
+    if (value & 0x00000010) == 0x00000010
+      raw_string += 'READ_PROPERTY|'
+    end
+    if (value & 0x00000008) == 0x00000008
+      raw_string += 'SELF_WRITE|'
+    end
+    if (value & 0x00000004) == 0x00000004
+      raw_string += 'LIST_CHILDREN|'
+    end
+    if (value & 0x00000002) == 0x00000002
+      raw_string += 'DELETE_CHILD|'
+    end
+    if (value & 0x00000001) == 0x00000001
+      raw_string += 'CREATE_CHILD|'
+    end
+
+    raw_string.chomp('|')
+  end
+
+  def parse_dacl_or_sacl(acl)
+    flag_allowed_to_enroll = nil
+    allowed_sids = []
+    acl.aces.each do |ace|
+      ace_header = ace[:header]
+      ace_body = ace[:body]
```

Comment:
Hmm so breakpointing on this shows that `ace[:body].class` is a `BinData::Choice` structure. However you are correct in that if the structure cannot be identified we will return a string instead of a hash when fetching the body contents. Whats weird is that despite being a string the object will still be identified as a `BinData::Choice` object type:


```
> ace_body
"0\x01\x00\x00\x01\x00\x00\x00h\xC9\x10\x0E\xFBx\xD2\x11\x90\xD4\x00\xC0Oy\xDCU\x01\x05\x00\x00\x00\x00\x00\x05\x15\x00\x00\x00k\x91\x19\xC4\x89\n" + "\xA3i\xBBPR\xC2\x00\x02\x00\x00"
> ace[:body].class
BinData::Choice
> ace_body.class
BinData::Choice
```

So we cannot use the class to determine what the type of the data is as I was expecting to do.

The one way I have found to see if its valid or not is to simply access one of the fields, for example `ace_body['access_mask']`. If it is the valid type, then it will return that data. Otherwise it will return `nil`. However this still assumes that the data will always contain an access mask or the similar field. That being said given that if we don't have this field the data is invalid anyway I think this is a fair assumption to make.